ouroboros-consensus/test-consensus/Test/Dynamic/Util/TestConfig.hs

```diff
@@ -0,0 +1,482 @@
+    , p'             <- andId shrinkNodeJoinPlan truncateP
+    , top'           <- andId shrinkNodeTopology truncateTop
+    ]
+
+-- | Shrink, including the number of slots but not number of nodes
+--
+shrinkTestConfigSlotsOnly :: TestConfig blk-> [TestConfig blk]
+shrinkTestConfigSlotsOnly TestConfig
+  { adhocTestConfig
+  , numCoreNodes
+  , numSlots
+  , nodeJoinPlan
+  , nodeTopology
+  } =
+    dropId $
+    [ TestConfig
+        { adhocTestConfig
+        , nodeJoinPlan    = p'
+        , nodeTopology    = top'
+        , numCoreNodes
+        , numSlots        = t'
+        }
+    | t'             <- andId shrink numSlots
+    , let truncateP   = truncateNodeJoinPlan nodeJoinPlan numCoreNodes (numSlots, t')
+    , let truncateTop = truncateNodeTopology nodeTopology numCoreNodes
+    , p'             <- andId shrinkNodeJoinPlan truncateP
+    , top'           <- andId shrinkNodeTopology truncateTop
+    ]
+
+truncateNodeJoinPlan ::
+    NodeJoinPlan -> NumCoreNodes -> (NumSlots, NumSlots) -> NodeJoinPlan
+truncateNodeJoinPlan
+  (NodeJoinPlan m) (NumCoreNodes n') (NumSlots t, NumSlots t') =
+    NodeJoinPlan $
+    -- scale by t' / t
+    Map.map (\(SlotNo i) -> SlotNo $ (i * toEnum t') `div` toEnum t) $
+    -- discard discarded nodes
+    Map.filterWithKey (\(CoreNodeId nid) _ -> nid < n') $
+    m
+
+truncateNodeTopology :: NodeTopology -> NumCoreNodes -> NodeTopology
+truncateNodeTopology (NodeTopology m) (NumCoreNodes n') =
+    NodeTopology $ Map.filterWithKey (\(CoreNodeId i) _ -> i < n') m
+
+{-------------------------------------------------------------------------------
+  MockBFT test
+-------------------------------------------------------------------------------}
+
+type BlockMockBFT = SimpleBftBlock SimpleMockCrypto BftMockCrypto
+
+instance TestableBlock BlockMockBFT where
+    type AdhocTestConfig BlockMockBFT = SecurityParam
+
+    genTestConfig = join $
+      defaultGenTestConfig (SecurityParam 5)
+        <$> arbitrary
+        <*> arbitrary
+
+    shrinkTestConfig = shrinkTestConfigFreely
+
+    adhocInconsistenciesTestConfig _ = []
+
+{-------------------------------------------------------------------------------
+  MockPraos test
+-------------------------------------------------------------------------------}
+
+type BlockMockPraos = SimplePraosBlock SimpleMockCrypto PraosMockCrypto
+
+defaultPraosParams :: PraosParams
+defaultPraosParams = PraosParams
+    { praosSecurityParam = SecurityParam 5
+    , praosSlotsPerEpoch = 3
+    , praosLeaderF       = 0.5
+    , praosLifetimeKES   = 1000000
+    }
+
+threePraosEpochs :: PraosParams -> NumSlots
+threePraosEpochs PraosParams
+  { praosSecurityParam = paramK
+  , praosSlotsPerEpoch
+  } =
+    NumSlots $ fromEnum $
+    3 * maxRollbacks paramK * praosSlotsPerEpoch
+
+instance TestableBlock BlockMockPraos where
+    type AdhocTestConfig BlockMockPraos = PraosParams
+
+    genTestConfig = do
+        let params   = defaultPraosParams
```

Comment:
Maybe `SecurityParam 5` instead of `defaultPraosParams` is more explicit? Or do you relying `defaultPraosParams` elsewhere?

Reply:
`PraosParams` is a record type, with more than just `SecurityParam`.

Reply:
I minimized this PR, so this content will be in a different PR. Resolving this conversation.